packages/proto-signing/src/signing.spec.ts

```diff
@@ -0,0 +1,359 @@
+        "cosmos",
+        msgSendFields.toAddress,
+      )} 1234567ucosm`,
+    );
+    console.log("");
+    console.log("Signed tx with local account:");
+    console.log("simd keys add -i testgen");
+    console.log(faucet.mnemonic);
+    console.log("");
+    console.log("simd keys show -a testgen");
+    console.log(`Should give ${address}`);
+    console.log("");
+    console.log("This doesn't seem to work.... but should sign without broadcast");
+    console.log(
+      `simd tx bank send --offline --chain-id ${chainId} -a 1 testgen ${Bech32.encode(
+        "cosmos",
+        msgSendFields.toAddress,
+      )} 1234567ucosm`,
+    );
+  });
+
+  fit("correctly parses and generates test vectors", async () => {
+    const myRegistry = new Registry();
+    const wallet = await Secp256k1Wallet.fromMnemonic(faucet.mnemonic);
+    const [{ address, pubkey: pubkeyBytes }] = await wallet.getAccounts();
+    const publicKey = PublicKey.create({
+      secp256k1: pubkeyBytes,
+    });
+
+    const parsedTestTx = Tx.decode(fromHex(signedTxBytesSeq0));
+    console.log(JSON.stringify(parsedTestTx));
+    expect(parsedTestTx.signatures.length).toEqual(1);
+    expect(parsedTestTx.authInfo?.signerInfos?.length).toEqual(1);
+    expect(parsedTestTx.authInfo?.signerInfos![0].publicKey!.secp256k1).toEqual(pubkeyBytes);
+    expect(parsedTestTx.authInfo?.signerInfos![0].modeInfo!.single!.mode).toEqual(1);
+    expect(parsedTestTx.authInfo?.fee!.amount).toEqual([]);
+    expect(parsedTestTx.authInfo?.fee!.gasLimit!.toString()).toEqual("200000");
+    expect(parsedTestTx.body?.extensionOptions).toEqual([]);
+    expect(parsedTestTx.body?.nonCriticalExtensionOptions).toEqual([]);
+    expect(parsedTestTx.body!.messages!.length).toEqual(1);
+
+    const parsedTestTxMsg = defaultRegistry.decode({
+      typeUrl: parsedTestTx.body!.messages![0].type_url!,
+      value: parsedTestTx.body!.messages![0].value!,
+    });
+    expect(parsedTestTxMsg.from_address).toEqual(Bech32.decode(address).data);
+    expect(parsedTestTxMsg.to_address).toEqual(
+      Uint8Array.from([1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20]),
+    );
+    expect(parsedTestTxMsg.amount.length).toEqual(1);
+    expect(parsedTestTxMsg.amount[0].denom).toEqual("ucosm");
+    expect(parsedTestTxMsg.amount[0].amount).toEqual("1234567");
+
+    const txBodyFields: TxBodyValue = {
+      messages: [
+        {
+          typeUrl: "/cosmos.bank.MsgSend",
+          value: {
+            fromAddress: parsedTestTxMsg.from_address,
+            toAddress: parsedTestTxMsg.to_address,
+            amount: [
+              {
+                denom: parsedTestTxMsg.amount[0].denom,
+                amount: parsedTestTxMsg.amount[0].amount,
+              },
+            ],
+          },
+        },
+      ],
+    };
+    const txBodyBytes = myRegistry.encode({
+      typeUrl: "/cosmos.tx.TxBody",
+      value: txBodyFields,
+    });
```

Comment:
Does this do the two step serialization of messages? A message is added to TxBody as raw bytes

Reply:
Yes: https://github.com/CosmWasm/cosmjs/blob/master/packages/proto-signing/src/registry.ts#L65-L90